File: py110/small_problems_110_119/advanced/3.py
# https://launchschool.com/exercises/564614de?track=python


'''
input: matrix
output: new matrix rotated 90 degrees
rules:

row = 2
col = 4

0 1 2 3

3 7 4 2 | 0
5 1 0 8 | 1


row = 4
col = 2

5 3
1 7
0 4
8 2

[[5, 3]]


0 1 2

1 5 8 | 0
4 7 2 | 1
3 9 6 | 2


3 4 1
9 7 5
6 2 8 

for c in col -> 3
    sub_list = [1, 4, 3]
    for r in row -> 3
    c = 0
    r = 2
    sub_list.append(matrix[r][c]) # reverse


Algorithm:
1. Create empty matrix
2. iterate from colum
3. initialize empty sublist
4. itereate through row
5. append reversed(row and column. )
6. return list


'''

matrix1 = [
    [1, 5, 8],
    [4, 7, 2],
    [3, 9, 6],
]

matrix2 = [
    [3, 7, 4, 2],
    [5, 1, 0, 8],
]

def rotate90(matrix):
    row, col = len(matrix), len(matrix[0])
    new_matrix = []

    for c in range(col):
        sub_list = []
        for r in range(row):
            sub_list.append(matrix[r][c])
        new_matrix.append(sub_list[::-1])
    
    return new_matrix


new_matrix1 = rotate90(matrix1)
new_matrix2 = rotate90(matrix2)
new_matrix3 = rotate90(rotate90(rotate90(rotate90(matrix2))))

# These examples should all print True
print(new_matrix1 == [[3, 4, 1], [9, 7, 5], [6, 2, 8]])
print(new_matrix2 == [[5, 3], [1, 7], [0, 4], [8, 2]])
print(new_matrix3 == matrix2)

File: ls215/smallProblems/adv/3.js
'use strict';

/*
input: mxn matrix
output: return new matrix rotated 90 degrees
rules:
1. Do not mutate input.

  0 1 2
0 1 5 8
1 4 7 2
2 3 9 6



  0 1 2
0 3 4 1
1 9 7 5
2 6 2 8


1 5 8
4 7 2
3 9 6

1 4 3
5 7 9
8 2 6

3 4 1
9 7 5
6 2 8

3 4 1
9 7 5
6 2 8



0,0 -> 0, 2
0,1 -> 1, 2
0,2 -> 2, 2

1,0 -> 0,1
1,1 -> 1,1
1,2 -> 2,1

2,0 -> 0,0
2,1 -> 1,0
2,2 -> 2,0

row = 2
col = 4

3 7 4 2
5 1 0 8

5 3
1 7
0 4
8 2

row ->4
col -> 2

row = 2
col = 1

[3, 5]
[7, 1]
[4, 0]

matrix[col][row]
matrix[1][2]


Q:
1.

DS:

A:
1. create result = []

Transpose
2. Itereate 0 to row to col length
3. create subarray
4. iterate 0 to col to row length
5. access matrix[col][row] and push to subarray
6. push to result


Reverse
1. iterate through result
2. reverse each row
3. return result

*/



const matrix1 = [
  [1, 5, 8],
  [4, 7, 2],
  [3, 9, 6],
];

const matrix2 = [
  [3, 7, 4, 2],
  [5, 1, 0, 8],
];

function rotate90(matrix) {
  return reverseArray(transpose(matrix));
}

function transpose(array) {
  let result = [];

  for (let row = 0; row < array[0].length; row += 1) {
    let subarray = [];
    for (let col = 0; col < array.length; col += 1) {
      subarray.push(array[col][row]);
    }

    result.push(subarray);
  }

  return result;
}

function reverseArray(array){
  return array.map(subarray => subarray.reverse());
}

const newMatrix1 = rotate90(matrix1);
const newMatrix2 = rotate90(matrix2);
const newMatrix3 = rotate90(rotate90(rotate90(rotate90(matrix2))));

console.log(newMatrix1);      // [[3, 4, 1], [9, 7, 5], [6, 2, 8]]
console.log(newMatrix2);      // [[5, 3], [1, 7], [0, 4], [8, 2]]
console.log(newMatrix3);      // `matrix2` --> [[3, 7, 4, 2], [5, 1, 0, 8]]